### DISCUSSION TITLE
Real Time decoding

### DISCUSSION THREAD

**Discussion by pratikkumar018** [espnet/espnet#3753] [Q&A]
We are using conformer based ASR models as shown in https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml.
and we are decoding using decode.yaml. 
We want to use this model for real-time decoding like for streaming purposes. Is there any way it is feasible in espnet1/espnet2?

Thanks for your response.

**Comment by sw005320** [comment]
We've made a demo
https://github.com/espnet/notebook/blob/master/espnet2_streaming_asr_demo.ipynb
This is only for Mandarin, but @D-Keqi will add English and Japanese models.
We may need more documentation and we welcome any input.

**Reply by pratikkumar018** [reply]
Can we use our own trained model for some other language?

**Reply by D-Keqi** [reply]
Sure! @pratikkumar018 But you may need to make sure that you have the latest version of ESPnet, and there is an example: https://github.com/espnet/espnet/blob/master/egs2/aishell/asr1/run_streaming.sh

**Reply by pratikkumar018** [reply]
Thanks for your response. We are able to run it using run_streaming.sh & asr_streaming.sh. But we are  getting decoding in traditional way where we have train/dev/eval folders in data & then we run Stage 3: Format wav.scp  stage to get the dump folder ready & then all those subsequent stages. I am not sure how we will do this in case of online/streaming. 
Can you please guide us how to do in case of streaming. Do I need to run all those stages again & again for all the wav files generated.
My current understanding is:


> Do this in loop:

    

>       1) Save the audio file in .wav format.
>       2) Run stage 3 Format wav.scp
>       3) Run stage 4.
>       4) Run decoding. 
> 

Is my understanding right or is there any other better way to do this?

If possible please provide any running example.

**Reply by D-Keqi** [reply]
I think you just need to retrain (stage 10 in espnet2 ) a model and decode (stage 11) online following the run_streaming.sh.
It may not work if you want to directly use the model following https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml
for streaming. For streaming, you need a contextual_block_conformer based encoder not a vanilla conformer encoder.

**Reply by pratikkumar018** [reply]
Thanks. I understand we need to train the model again here. (Running stage 10).
My question is once I have a trained asr_streaming model, how to do continuous streaming decode ? 
Should I keep updating the wav file or generate new ones ? and once I have the new/updated wav file do I need to do feature extraction again(stage 3) and then do the decoding.. 

Do I need to do all these things in a loop? My basic doubt is how will espnet know that the wav file has changed. It will run once & exit  unless we are running this in a loop.

**Reply by D-Keqi** [reply]
You do not need to do feature extraction again in espnet2 if you choose feats_type=raw.

**Reply by pratikkumar018** [reply]
Thanks I am able to replicate this & run decoding as shown in https://github.com/espnet/notebook/blob/master/espnet2_streaming_asr_demo.ipynb.

Now my only issue is as I keep on speaking my wav file is changing, but that is not happening in decoded text. It is only reading the wav file once & is transcribing only till that part. Change is my wav file is not getting reflected in realtime. I want the file to be continuously read . Is it possible?

**Reply by D-Keqi** [reply]
I think it works, let's say we have a speech A=[[chunk a1], [chunk a2], [chunk a3]], speech b=[[chunk b1], [chunk b2], [chunk b3]], I think we can just concatenate them into [[chunk a1], [chunk a2], [chunk a3],[chunk b1], [chunk b2], [chunk b3]], and then feed it into the model chunk by chunk for online deocoding. Maybe you can still be inspired by my inplementation https://github.com/espnet/notebook/blob/master/espnet2_streaming_asr_demo.ipynb

**Reply by pratikkumar018** [reply]
Got it. I changed the recognize function a bit and now it is able to detect the changes in file. I am keeping a track of last position till where it has read & starting from there is subsequent loops. I am breaking the loop if there is nothing changed in the read pointer . My implementation(** represents my changes):
```
def recognize(wavfile1):
    **pointer=0**
    while True:
        with wave.open(wavfile1, 'rb') as wavfile:
            **wavfile.setpos(pointer)**
            ch=wavfile.getnchannels()
            bits=wavfile.getsampwidth()
            rate=wavfile.getframerate()
            nframes=wavfile.getnframes()
            buf = wavfile.readframes(-1)
            data=np.frombuffer(buf, dtype='int16')
            speech = data.astype(np.float16)/32767.0
            **newpointer=wavfile.tell()
            if(newpointer==pointer):
                break
            else:
                pointer=newpointer**
```

**Reply by D-Keqi** [reply]
Very cool!

**Reply by pratikkumar018** [reply]
Hi, 
Is this the latest or any further development has happened in the streaming decoding ? Thanks for you reply.

**Reply by D-Keqi** [reply]
> Hi, Is this the latest or any further development has happened in the streaming decoding ? Thanks for you reply.

Hi, some further development can be found from the demo https://github.com/espnet/notebook/blob/master/espnet2_streaming_asr_demo.ipynb, and we also extend it to streaming SLU and ST tasks.

**Reply by pratikkumar018** [reply]
Thanks for the info.
